### ISSUE TITLE
Support "clear console"

### ISSUE THREAD

**Issue by roblourens** [Microsoft/vscode-debugadapter-node#185]
https://github.com/Microsoft/vscode/issues/50461#issuecomment-394158879

Chrome/node support `console.clear()` to clear the console. We could support this in the debug adapter protocol.

**Comment by gregg-miskelly** [comment]
@roblourens has does that work?

**Comment by mickaelistria** [comment]
I don't think it fits in the protocol. I imagine each client should implement it's own console clear. After all, a console is just a bunch of text, is this feature of debugging really different than any other kind of console one can show? I don't think so. And should any console provide a "clear" in any case? why not.
It's just a client-specific feature. FWIW, Eclipse IDE console allows to clear any console already, on a pure text approach. This works great, doesn't expect a provider for clearing and is reusable by any console user.

**Comment by roblourens** [comment]
When `console.clear()` is called in Node/Chrome, the runtime sends `{"method":"Runtime.consoleAPICalled","params":{"type":"clear" ...`. I don't know whether other runtimes do something similar or if it's more typical for the client to intercept some repl command to clear the console.

This could also be done as a one-off just for the node/chrome adapters in VS Code, using a custom event to invoke the 'clear repl' command in VS Code.

> FWIW, Eclipse IDE console allows to clear any console already, on a pure text approach.

What do you mean, on a pure text approach?

**Comment by mickaelistria** [comment]
Ah OK, so you meant clearing as runtime sending an event to clear the console.
I have to admit I find it a bit strange that a runtime may decide to clear a console, as a console is usually just a set of accumulated text file.
Also, similarly to Test events discussed in other stuff; isn't advanced console management (ie making it event-based rather than plain text display) an orthogonal stuff to the actual debug?

> What do you mean, on a pure text approach?

I mean the generic Eclipse IDE console, that the DAP support uses, view as a "clear" action button and when clicked, it just wipes out the existing text.

**Comment by roblourens** [comment]
I see, vscode has a similar button. I think some users are used to typing `clear()` (same as `console.clear`) or clearing the console from scripts.

From the C# side, I see a similar function: https://msdn.microsoft.com/en-us/library/system.console.clear.aspx?f=255&MSPPError=-2147217396

**Comment by weinand** [comment]
I'm a bit reluctant to add this to the protocol.

The Chrome/Node DA could send a 'clear' ANSI Escape Sequence (ESC[2J) to the debug console.

VS Code honours this already:
https://github.com/Microsoft/vscode/blob/78d7d64c5d16b8d89d1dd2b0cc7fcbe966feb48b/src/vs/workbench/parts/debug/electron-browser/debugService.ts#L671

**Comment by roblourens** [comment]
That code is checking `[2J` **without** the escape character! If the adapter sends literally `[2J` it will clear the console. The node/chrome adapter adds a newline to 	console.log messages so we can't hit it with `console.log('2[J')` fortunately :)

**Comment by nebosite** [comment]
I found this thread because I'm trying to us VSCode to debug a console window that clears the screen and uses cursor position to do simple graphics.  (Very simple learning project for HS kids.)  

So, multiple problems:

1. Both Console.Clear() and Console.SetCursorPostion()  crash in debug dur to an IOException "The handle is invalid"
2. For kicks, I tried replacing Console.Clear() with the following code and it did not clear the debug console: Console.Write($"\x1b]2J");

**Comment by weinand** [comment]
@nebosite and you are writing a console program in C# that sends data to the console?

**Comment by nebosite** [comment]
@weinand - yes, just a very simple console app:

```
class Program
    {
        static void Main(string[] args)
        {
            Console.Write($"\x1b]2J");
            Console.SetCursorPosition(5,5);

            //Console.Clear();
        }
    }
```

**Comment by weinand** [comment]
You are in the wrong repository. This place is for a node.js based support library to write debuggers for VS Code. It has nothing todo with your problem.

I assume you are using the C# extension (which includes the debugger). I suggest to file an issue against the C# extension.

**Comment by nebosite** [comment]
OK, thanks

**Comment by gregg-miskelly** [comment]
@nebosite If you are trying to use the console for something other than simple output, you want to set `"console": "integratedTerminal"` (or `externalTerminal` if you prefer) in your launch.json.
